plugin/core/promise.py

```diff
@@ -0,0 +1,194 @@
+            # the promise with provided value and passes the value to the next
+            # chained promise.
+            resolve(111)  # Can be invoked asynchronously.
+
+        def process_value(value):
+            assert value === 111
+
+        Promise(do_work_async).then(process_value)
+
+      * Returning Promise from chained promise:
+
+        def do_work_async_1(resolve):
+            # Compute value asynchronously.
+            resolve(111)
+
+        def do_work_async_2(resolve):
+            # Compute value asynchronously.
+            resolve(222)
+
+        def do_more_work_async(value):
+            # Do more work with the value asynchronously. For the sake of this
+            # example, we don't use 'value' for anything.
+            assert value === 111
+            return Promise(do_work_async_2)
+
+        def process_value(value):
+            assert value === 222
+
+        Promise(do_work_async_1).then(do_more_work_async).then(process_value)
+    """
+
+    @classmethod
+    def resolve(cls, resolve_value=None):
+        """Immediately resolves a Promise.
+
+        Convenience function for creating a Promise that gets immediately
+        resolved with the specified value.
+
+        Arguments:
+            resolve_value: The value to resolve the promise with.
+        """
+        def executor(resolve_fn):
+            return resolve_fn(resolve_value)
+
+        return cls(executor)
+
+    @classmethod
+    def all(cls, promises: List['Promise']) -> 'Promise':
+        """
+        Takes a list of promises and returns a Promise that gets resolved when all promises
+        gets resolved.
+
+        :param      promises: The list of promises
+
+        :returns:   A promise that gets resolved when all passed promises gets resolved.
+                    Gets passed a list with all resolved values.
+        """
+        def handler(resolve):
+            def recheck_resolve_status(_: Any):
+                # We're being called from a Promise that is holding a lock so don't try to use
+                # any methods that would try to acquire it.
+                if all(p.resolved for p in promises):
+                    resolve([p.value for p in promises])
+            for p in promises:
+                if not isinstance(p, Promise):
+                    raise Exception('Value is not a Promise')
+                p.then(recheck_resolve_status)
+
+        if promises:
+            return Promise(handler)
+        return Promise.resolve()
+
+    def __init__(self, executor):
+        """Initialize Promise object.
+
+        Arguments:
+            executor: A function that is executed immediately by this Promise.
+            It gets passed a "resolve" function. The "resolve" function, when
+            called, resolves the Promise with the value passed to it.
+        """
+        self.value = None
+        self.resolved = False
+        self.mutex = threading.Lock()
```

Comment:
If we can make it so that the promise's callbacks are always run on ST's worker thread, then I think we don't need this lock.

Reply:
> what is the advantage of having this class over `concurrent.futures.Future`

I wasn't aware of `concurrent.futures.Future` and have no experience with it. I would have to read up on it but possibly we could just use that instead.

> If we can make it so that the promise's callbacks are always run on ST's worker thread, then I think we don't need this lock.

Things get complicated quickly.
 - if we run everything on the worker thread then on which thread do we call back once the promise is resolved? We can't jump back to the same thread where the promise was created. At least not that I know how to do that.
 - why introduce extra thread jumps if they are not always needed?
 - on adding a `then` callback, the promise would have to potentially internally jump on a different thread just to check if a promise is resolved. That would be less performant and more complicated.

Reply:
I know that concurrent.futures.Future is not monadic while this implementation does seem to be monadic. It's probably best to go ahead with the implementation presented in the diff.